### PR TITLE
FLAML classifier metrics

### PR DESCRIPTION
**Pull request by prithvikannan**
Signed-off-by: Prithvi Kannan <prithvi.kannan@databricks.com>

## Related Issues/PRs

https://github.com/mlflow/mlflow/pull/6959

## What changes are proposed in this pull request?

Adding classifier metrics to the `_MLFLOW_TO_FLAML_METRICS` mapping. This unblocks users who want to use automl training for classification problems.

FLAML only supports `f1_score` and `accuracy` natively. We have built adapters for `recall_score` and `precision_score`, since we support those as primary metrics.

## How is this patch tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests (describe details, including test results, below)

E2E test with the wine dataset classification example using automl/flaml in the train step.

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [x] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/recipes/steps/automl/flaml.py [hunk 1]
```diff
@@ -37,14 +44,18 @@ def get_estimator_and_best_params(
...
-        builtin_metrics = _get_regressor_metrics(y, y_pred, sample_weights=None)
+        builtin_metrics = (
+            _get_regressor_metrics(y, y_pred, sample_weights=None)
+            if task == "regression"
+            else _get_binary_classifier_metrics(y_true=y, y_pred=y_pred)
+        )
         res_df = pd.DataFrame()
         res_df["prediction"] = y_pred
-        res_df["target"] = y.values
+        res_df["target"] = y
```
Comment (prithvikannan) [inline mlflow/recipes/steps/automl/flaml.py:58]: This is a bug fix. `y` is already a numpy array

Comment (dbczumar) [reply mlflow/recipes/steps/automl/flaml.py:58]: Why doesn't this come up in regression & is there a test case that we need to add for this workflow?

File: mlflow/recipes/steps/automl/flaml.py [hunk 2]
```diff
@@ -19,8 +22,12 @@
     "root_mean_squared_error": "rmse",
     "r2_score": "r2",
     "mean_absolute_percentage_error": "mape",
+    "f1_score": "f1",
+    "accuracy_score": "accuracy",
 }
 
+_MLFLOW_TO_SKLEARN_METRICS = ["recall_score", "precision_score"]
```
Comment (dbczumar) [inline mlflow/recipes/steps/automl/flaml.py:29]: `_TO_` implies a dictionary mapping. Should we rename this?

**Review by dbczumar — APPROVED** [review]
LGTM once a couple tiny comments are addressed & we've added tests. Thanks so much, @prithvikannan !

cc @BenWilson2 can we cherry-pick this into the MLflow 2.0 release branch?

File: mlflow/recipes/steps/automl/flaml.py (hunk 1 above)
Comment (prithvikannan) [reply mlflow/recipes/steps/automl/flaml.py:59]: Looks like our current test coverage does not include a custom metric for automl. I'll add one.

https://github.com/mlflow/mlflow/blob/902c103b64a8e3c48ed081aab09c46937a4d3389/tests/recipes/test_train_step.py#L242-L246

**Review by mshtelma — APPROVED** [review]
Thanks, Prithvi! LGTM!

File: mlflow/recipes/steps/automl/flaml.py (hunk 1 above)
Comment (prithvikannan) [reply mlflow/recipes/steps/automl/flaml.py]: EDIT: I was wrong. For regression automl we do test for a custom error function.
https://github.com/mlflow/mlflow/blob/902c103b64a8e3c48ed081aab09c46937a4d3389/tests/recipes/test_train_step.py#L517-L525

**Comment by jinzhang21** [comment]
Many thanks for the quick fix, @prithvikannan . Do you mind filing a follow up ticket / PR to add some unit test to ensure classification continue to work for AutoML?

**Pull request merged**